From 3411d1edd25c7d82607b62c4c83755b598e921ef Mon Sep 17 00:00:00 2001
From: Cheick Traore <cheick.traore@foss.st.com>
Date: Mon, 16 Sep 2024 15:39:44 +0200
Subject: [PATCH 1524/1695] drivers: rework counter API

Prepare to replace alarm by event. Add an enum containing an overflow
and threshold event of the counter.
Add a threshold and ceiling set in counter ops to independently set them
as they are specific to the counter and not the event.
Each event have its own callback stored in the struct counter_event.
Canceling counter event cancels all the events.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ea5cb048fa9bfe4cc34415df72e803b71aec50c5 drivers: rework counter API]
Change-Id: I527a040355d8b406ffb9307fee619ff1c1b38086
Signed-off-by: Cheick Traore <cheick.traore@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/403986
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
---
 core/drivers/counter/counter.c | 152 ++++++++++++++++++++++++++++++++-
 core/include/drivers/counter.h |  94 ++++++++++++++++++++
 2 files changed, 245 insertions(+), 1 deletion(-)

diff --git a/core/drivers/counter/counter.c b/core/drivers/counter/counter.c
index a2c7b6ad5..4e8c9882a 100644
--- a/core/drivers/counter/counter.c
+++ b/core/drivers/counter/counter.c
@@ -36,7 +36,7 @@ TEE_Result counter_start(struct counter_device *counter, void *config)
 	return counter->ops->start(counter, config);
 }
 
-/*
+/**
  * some counter device could not be stopped
  * return TEE_SUCCESS if no stop ops
  */
@@ -108,6 +108,156 @@ TEE_Result counter_cancel_alarm(struct counter_device *counter)
 	return res;
 }
 
+TEE_Result counter_set_threshold(struct counter_device *counter,
+				 unsigned int ticks)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(counter && counter->ops);
+
+	if (!counter->ops->set_threshold)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (ticks > counter->ceiling)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = counter->ops->set_threshold(counter, ticks);
+	if (res)
+		return res;
+
+	counter->threshold = ticks;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result counter_set_ceiling(struct counter_device *counter,
+			       unsigned int ceiling)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(counter && counter->ops);
+
+	if (!counter->ops->set_ceiling)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (ceiling > counter->max_ticks || ceiling < counter->threshold)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = counter->ops->set_ceiling(counter, ceiling);
+	if (res)
+		return res;
+
+	counter->ceiling = ceiling;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result counter_call_callback(struct counter_device *counter,
+				 enum counter_event_type event_type)
+{
+	struct counter_event *event = &counter->events[event_type];
+
+	if (!event->is_enabled)
+		return TEE_ERROR_BAD_STATE;
+
+	if (!event->callback)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	event->callback(event->priv, event_type);
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * counter_set_threshold and counter_set_ceiling must be called before enabling
+ * an event if needed to adjust threshold and ceiling.
+ */
+TEE_Result counter_enable_event(struct counter_device *counter,
+				enum counter_event_type event_type,
+				counter_event_cb_t callback, void *priv)
+{
+	struct counter_event *event = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(counter && counter->ops);
+
+	if (!counter->ops->enable_event)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	switch (event_type) {
+	case COUNTER_EVENT_OVERFLOW:
+	case COUNTER_EVENT_THRESHOLD:
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	if (!callback)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	event = &counter->events[event_type];
+	event->is_enabled = true;
+	event->callback = callback;
+	event->priv = priv;
+	res = counter->ops->enable_event(counter, event_type);
+	if (res) {
+		event->is_enabled = false;
+		event->callback = NULL;
+		event->priv = NULL;
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result counter_disable_event(struct counter_device *counter,
+				 enum counter_event_type event_type)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(counter && counter->ops);
+
+	if (!counter->ops->disable_event)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	switch (event_type) {
+	case COUNTER_EVENT_OVERFLOW:
+	case COUNTER_EVENT_THRESHOLD:
+		break;
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+	if (!counter->events[event_type].is_enabled)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = counter->ops->disable_event(counter, event_type);
+	if (res)
+		return res;
+
+	counter->events[event_type].is_enabled = false;
+	counter->events[event_type].callback = NULL;
+	counter->events[event_type].priv = NULL;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result counter_disable_all_events(struct counter_device *counter)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(counter && counter->ops);
+
+	for (int i = 0; i < COUNTER_NB_EVENT; i++) {
+		if (!counter->events[i].is_enabled)
+			continue;
+		res = counter_disable_event(counter, i);
+		if (res)
+			return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
 void counter_release_config(struct counter_device *counter,
 			    void *config)
 {
diff --git a/core/include/drivers/counter.h b/core/include/drivers/counter.h
index 3689756ff..5c7556c5f 100644
--- a/core/include/drivers/counter.h
+++ b/core/include/drivers/counter.h
@@ -27,6 +27,34 @@ struct alarm_cfg {
 	bool is_enabled;
 };
 
+enum counter_event_type {
+	/* Count value increased past ceiling */
+	COUNTER_EVENT_OVERFLOW,
+	/* Count value reached threshold */
+	COUNTER_EVENT_THRESHOLD,
+	COUNTER_NB_EVENT,
+};
+
+/**
+ * counter_event_cb_t - Typedef for counter event callback handler functions
+ * @priv: Event consumer private data
+ * @event: Id of the event reported
+ */
+typedef void (*counter_event_cb_t)(void *priv, enum counter_event_type event);
+
+/**
+ * @brief event structure.
+ *
+ * @param is_enabled True if event is enabled, false otherwise.
+ * @param callback Function called on event (cannot be NULL).
+ * @param priv Private data passed to the callback function.
+ */
+struct counter_event {
+	bool is_enabled;
+	counter_event_cb_t callback;
+	void *priv;
+};
+
 struct counter_device;
 
 struct counter_param {
@@ -34,12 +62,35 @@ struct counter_param {
 	size_t len;
 };
 
+/**
+ * struct counter_ops - Counter device operations
+ *
+ * @start: function to start the counter
+ * @stop: function to stop the counter
+ * @get_value: function to get the counter value
+ * @set_alarm: (deprecated)
+ * @cancel_alarm: (deprecated)
+ * @set_threshold: function to set the counter threshold
+ * @set_ceiling: function to set the counter ceiling
+ * @enable_event:function to enable an event on counter value
+ * @disable_event: function to clear all events
+ * @set_config: function to set a config in the counter
+ * @release_config: function to release a config from the counter
+ */
 struct counter_ops {
 	TEE_Result (*start)(struct counter_device *counter, void *config);
 	TEE_Result (*stop)(struct counter_device *counter);
 	TEE_Result (*get_value)(struct counter_device *counter, unsigned int *ticks);
 	TEE_Result (*set_alarm)(struct counter_device *counter);
 	TEE_Result (*cancel_alarm)(struct counter_device *counter);
+	TEE_Result (*set_threshold)(struct counter_device *counter,
+				    unsigned int ticks);
+	TEE_Result (*set_ceiling)(struct counter_device *counter,
+				  unsigned int ticks);
+	TEE_Result (*enable_event)(struct counter_device *counter,
+				   enum counter_event_type event_type);
+	TEE_Result (*disable_event)(struct counter_device *counter,
+				    enum counter_event_type event_type);
 	TEE_Result (*set_config)(struct counter_device *counter,
 				 const void *param,
 				 int len, void **config);
@@ -55,7 +106,10 @@ struct counter_ops {
  * @param is_used True if counter is used (exclusive consumer).
  * @param ops Operation table of the counter.
  * @param alarm Alarm configuration of the counter.
+ * @param events List of event of the counter.
  * @param max_ticks Tick max value supported by the counter.
+ * @param threshold Value reached on threshold event (COUNTER_EVENT_THRESHOLD).
+ * @param ceiling Value reached on overflow event (COUNTER_EVENT_OVERFLOW).
  * @param priv Optional private data supplied by driver.
  */
 struct counter_device {
@@ -65,7 +119,10 @@ struct counter_device {
 	bool is_used;
 	const struct counter_ops *ops;
 	struct alarm_cfg alarm;
+	struct counter_event events[COUNTER_NB_EVENT];
 	unsigned int max_ticks;
+	unsigned int threshold;
+	unsigned int ceiling;
 	void *priv;
 };
 
@@ -94,6 +151,43 @@ TEE_Result counter_set_alarm(struct counter_device *counter);
  */
 TEE_Result counter_cancel_alarm(struct counter_device *counter);
 
+/**
+ * @brief Set counter threshold.
+ */
+TEE_Result counter_set_threshold(struct counter_device *counter,
+				 unsigned int ticks);
+
+/**
+ * @brief Set counter ceiling.
+ */
+TEE_Result counter_set_ceiling(struct counter_device *counter,
+			       unsigned int ceiling);
+
+/**
+ * @brief Call of the registered callback function of the event.
+ */
+TEE_Result counter_call_callback(struct counter_device *counter,
+				 enum counter_event_type event_type);
+
+/**
+ * @brief Set an event and associate a callback.
+ */
+TEE_Result counter_enable_event(struct counter_device *counter,
+				enum counter_event_type event_type,
+				counter_event_cb_t callback,
+				void *priv);
+
+/**
+ * @brief Disable an event.
+ */
+TEE_Result counter_disable_event(struct counter_device *counter,
+				 enum counter_event_type event_type);
+
+/**
+ * @brief Disable all events.
+ */
+TEE_Result counter_disable_all_events(struct counter_device *counter);
+
 /**
  * @brief Release the counter configuration.
  */
-- 
2.47.2

